notebook/tests/selenium/utils.py

```diff
@@ -382,3 +386,77 @@ def trigger_keystrokes(browser, *keys):
             ac.perform()
         else:              # single key stroke. Check if modifier eg. "up"
             browser.send_keys(getattr(Keys, keys[0].upper(), keys[0]))
+
+def validate_dualmode_state(notebook, mode, index):
+    '''Validate the entire dual mode state of the notebook.  
+    Make sure no more than one cell is selected, focused, in edit mode, etc...
+
+    '''
+    def is_only_cell_edit(index):
+        JS = 'return Jupyter.notebook.get_cells().map(function(c) {return c.mode;})'
+        cells_mode = notebook.browser.execute_script(JS)
+        #None of the cells are in edit mode
+        if(index == None):
+            for mode in cells_mode:
+                if(mode == 'edit'):
+                    return False
+            return True
+        #Only the index cell is on edit mode
+        i = 0
+        for mode in cells_mode:
+            if(i == index):
+                if(mode != 'edit'):
+                    return False
+            else:
+                if(mode == 'edit'):
+                    return False
+            i += 1
+        return True
+
+    def is_focused_on(index):
+        JS = "return $('#notebook .CodeMirror-focused textarea').length;"
+        focused_cells = notebook.browser.execute_script(JS)
+        if(index == None):
+            return focused_cells == 0
+
+        assert focused_cells == 1
+
+        JS = "return $('#notebook .CodeMirror-focused textarea')[0];"
+        focused_cell = notebook.browser.execute_script(JS)
+        JS = "return IPython.notebook.get_cell(%s).code_mirror.getInputField()"%index
+        cell = notebook.browser.execute_script(JS)
+        assert focused_cell == cell
+
+
+    #general test
+    JS = "return IPython.keyboard_manager.mode;"
+    keyboard_mode = notebook.browser.execute_script(JS)
+    JS = "return IPython.notebook.mode;"
+    notebook_mode = notebook.browser.execute_script(JS)
+
+    #validate selected cell
+    JS = "return Jupyter.notebook.get_selected_cells_indices();"
+    cell_index = notebook.browser.execute_script(JS)
+    assert cell_index == [index] #only the index cell is selected
+
+    if(mode == 'command'):
+        #validate mode
+        assert mode == keyboard_mode #keyboard mode is correct
+
+        assert is_focused_on(None)#no focused cells
+
+        assert is_only_cell_edit(None)#no cells in edit mode
```

Comment:
Minor code style point: there should be one or two spaces between the code and the comment.

I also find comments are more readable with one space after the `#`, but that may be more of a personal preference.